src/firewall/functions.py

```diff
@@ -43,6 +43,43 @@ def fail(msg=None):
 ###############################################################################
 
 
+def addr_family(family, allow_unspec=False):
+    """Normalizes (and verifies) the address family. It returns the corresponding
+    socket.AF_INET, socket.AF_INET6, or socket.AF_UNSPE
+
+    @family: the family to normalize/verify.
+    @allow_unspec: whether socket.AF_UNSPEC and None is accepted as socket.AF_UNSPE
+    """
+    if family == socket.AF_INET or family == socket.AF_INET6:
+        return family
+    if family is None or family == socket.AF_UNSPEC:
+        if allow_unspec:
+            return socket.AF_UNSPEC
+    elif isinstance(family, str):
+        family = family.lower()
+        if family in ("ipv4", "4", "ip4", "inet4", "inet"):
+            return socket.AF_INET
+        if family in ("ipv6", "6", "ip6", "inet6"):
+            return socket.AF_INET6
+
+    # Family is not untrusted user-input, it's provider by the calling code.
+    # Providing an invalid family is a bug.
+    raise BugError.fail("not a valid IP address family")
```

Comment:
> I disagree. It's literally an error class. Put it in errors.py.

Errors.py contains the error codes from firewalld. It doesn't mean, that any Exception class should be there.

Optimally, there would be a most-low-level place, that contains code which only depends on Python stdlib and is not (very) firewalld specific. As such, it could contain `ipaddr_parse()` and also a `BugError`. That code is currently in `firewall.core.functions` (which I think is a bad name).

> I'll have to look at why firewall.core.icmp is being imported in functions.py.

`firewall.core.icmp` is low-level too. It literally contains a few strings with names of ICMP types. It's needed by `port_parse()` to get the names of the accepted ICMP types. If we accept that `port_parse()` is one of the most low level functions, then so must `firewall.core.icmp`.

Reply:
> IMO they're a distraction. Similar or related functions are adjacent.

I don't understand. Similar/related functions are adjacent, but these lines separate functions that are *not* related/similar.

They are visual clues, to easier see where a topic starts/ends and highlight those places.

Their purpose would be the opposite of distracting. Seems they fail in that.

Reply:
dropped `BugError.fail()` and always use `raise BugError`.